ls1intum/Artemis
# `Discussion:` Automatically sync postings between multiple users

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


Users that are on pages with postings, i.e. the course discussion page or certain lecture/exercise pages should see updates on postings (e.g., new posts, new answer posts, content changes, reactions etc.) immediately without having to reload the page.
They should have the feeling of live-communication.

### Description

Before, the shown posts where fetched from the database via REST calls that 
We integrated Websocket communication with the dedicated **channel/topic**.
- If a user is on a lecture page, the Metis Service has a subscription to the channel `/topic/metis/lectures/{lectureId}`
- If a user is on an exercise page, the Metis Service has a subscription to the channel `/topic/metis/exercises/{exerciseId}`
- If a user is on the course discussion page page, the Metis Service has a subscription to the channel `/topic/metis/courses/{courseId}`

There is only one subscription at a time.

Depending on the user action, POST/PUT/DELETE REST call to a server endpoint, a message is broadcasted on the according channel.
A message needs to have the channel/topic (depending on the post context), and a payload (`postDTO`).
These two components are built upon the affected post, note that also when an answer post changes, we account the original post as affected post.
The actual creation and sending of the websocket message is happening in `broadcastForPost` in the `PostingService.java` which is invoked in the according service methods of `PostService.java` and `AnswerPostService.java`
As an overview these action trigger following **websocket messages**:

| Action                                                | PostDTO.post | PostDTO.action |
| ---------------------------------- | ------------- | ----------------- |
| New Post A Created                         | Post A            | CREATE_POST |
| Post A Updated                                | Post A            | UPDATE_POST, if context changed: CREATE_POST (in new context), DELETE_POST (in old context) |
| Post A Deleted                                 | Post A            | DELETE_POST |
| New Answer B for Post A Created  | Post A            | CREATE_POST |
| Answer Post B for Post A Updated | Post A            | UPDATE_POST |
| Answer Post B for Post A  Deleted | Post A            | DELETE_POST |
| Reaction to Answer Post B for Post A Created | Post A            | UPDATE_POST |
| Reaction to Answer Post B for Post A Deleted | Post A            | UPDATE_POST |

### Steps for Testing

Prerequisites:
- 2 Different Users
- 1 Course with postings enabled, on TS e.g. https://artemistest.ase.in.tum.de/course-management/477/

1. Log in to Artemis
2. Open to windows with two different user logged, the users A and B can be at the same page or at different pages. A page should either be the course discussion overview or a certain exercise or lecture page as postings are only visible on those pages
3. With User A try different things:
- Add new post
- Add new answer post
- Edit post -> if user A is instructor, you should also try to change the context, i.e. move the post somewhere else
- Edit answer post
- Delete post
- Delete answer post
- React on post
- React on answer post
- Remove reaction on post
- Remove reaction on answer post
- Upvote a post
4. As User B check if the action of User A is reflected correctly without reloading the page, i.e., post/answerpost/reaction appears/disappears, order is changed, etc..

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage






metis.service.ts: 96%
post.service.ts: 100%
answer-post.service.ts: 100%
reaction.service.ts: 100%
page-discussion.component.ts: 84%
course-discussion.component.ts: 90%

### Screenshots


In this video you see how different users would experience interaction with postings using websocket communication (left is instructor, right is student):

https://user-images.githubusercontent.com/41366522/138892847-748edb36-49d2-44ec-9f10-09ff8d82c9e5.mov
